containenv/containenv/cli/commands/init.py

```diff
@@ -48,3 +77,152 @@ def __call__(self):
+        if not os.path.isdir(self.proj_path):
+            error = ('init must be called on a project directory but the '
+                     'argument it received was: {}\nwhich does not map to '
+                     'a path to an existing directory.'
+                     ''.format(self.metasource))
+            raise ProjectDirectoryDoesNotExist(error)
+
+    def check_initialization_state(self):
+        '''determine whether a target project exists'''
+        self._check_project_existence()
+        if os.path.isdir(os.path.join(self.proj_path, '.containenv')):
+            error = ('init can only be called once per project, but the {}\n'
+                     'project already has a ".containenv" directory.'
+                     ''.format(self.proj_path))
+            raise ProjectAlreadyInitialized(error)
+
+    def make_containenv_dir(self):
+        '''make PROJECT/.containenv'''
+        self.containenvdir = os.path.join(self.proj_path, '.containenv')
+        os.mkdir(self.containenvdir)
+
+    def _register_file(self, dirpath, fname, extension):
+        # First use file extensions
+        if extension in EXTENSION_MAP:
+            self.registered_dependency_catalog[
+                EXTENSION_MAP[extension][self.from_image]
+            ].add(os.path.join(dirpath, fname+extension))
+        # Then check against fnames
+        elif fname in FILENAME_MAP:
+            self.registered_dependency_catalog[
+                FILENAME_MAP[fname][self.from_image]
+            ].add(os.path.join(dirpath, fname+extension))
+        else:
+            self.unregistered_nodes.add(os.path.join(dirpath, fname+extension))
+
+    def _register_dirs(self, dirpath, dirn, dirs):
+        fullpath = os.path.join(dirpath, dirn)
+        if dirn in DIRNAME_MAP:
+            self.registered_dependency_catalog[
+                DIRNAME_MAP[dirn][self.from_image]].add(fullpath)
+            self.registered_directories.add(fullpath)
+        else:
+            self.unregistered_nodes.add(fullpath)
+
+    def _catalog_dependencies(self):
+        '''create a list of technologies used in the project'''
+        self.registered_directories = set()
+        self.registered_dependency_catalog = collections.defaultdict(set)
+        self.unregistered_nodes = set()
+        for dirpath, dirs, files in os.walk(self.proj_path):
+            if dirpath in self.registered_directories:
+                continue
+            for fname, extension in [os.path.splitext(f) for f in files]:
+                self._register_file(dirpath, fname, extension)
+            for dirn in dirs:
+                self._register_dirs(dirpath, dirn, dirs)
+        if not self.registered_dependency_catalog:
+            error = ('None of the nodes (files and directories) in this'
+                     ' project are registered with containenv.')
+            raise ProjectContainsNoRegisteredNodes(error)
+
+    def _render_config(self):
```

Comment:
Hmm. Yeah, maybe this needs to be somewhat invisible to the user after `init` has been run once. As part of `contain`?

I think I need you to explain how you see this being used in the common case before I can give a firm answer on this though, because I'm pretty sure I've been imagining it differently.